_spinnaker/automated_rollbacks.md

```diff
@@ -0,0 +1,147 @@
+
+![image](/assets/images/aut-rollbacks-infra-1.png)
+
+The first version deployed (`V001`) corresponds to nginx 1.14.0, while the second version (`V002`) corresponds to nginx 1.15.5-alpine.
+
+### Creating a rollback pipeline
+
+Then we’ll be creating a rollback pipeline that triggers automatically when `deploy-nginx` pipeline fails:
+
+![image](/assets/images/aut-rollbacks-rollpipeline-1.png)
+
+![image](/assets/images/aut-rollbacks-trigger.png)
+
+This pipeline just contains a single stage of type Undo Rollout (Manifest) with the following configuration:
+
+![image](/assets/images/aut-rollbacks-undo-rollout-1.png)
+
+This configuration is important, the fields `Namespace`, `Kind` and `Name` are the coordinates of the kubernetes deployment, and should match what is defined in the deployment manifest on the deploy pipeline. `Revisions Back` indicate how many deployment versions to rollback. In our historic running we currently have `V001` and `V002`. Each time a deployment occurs, this version is automatically increased. [Click here for detailed information on how Undo Rollout (Manifest) stage works](https://www.spinnaker.io/guides/user/kubernetes-v2/automated-rollbacks/#automated-rollbacks).
+
+### Testing rollback logic
+
+Now we’ll do a rollback test by changing the manifest providing an invalid value for the namespace to force a failure. In deploy-nginx we’ll be changing the namespace to an invalid value:
+
+![image](/assets/images/aut-rollbacks-deploy-manifest-invalid-1.png)
+
+We’ll run the pipeline again to see the rollback pipeline kick in automatically to rollback one version of the deployment:
+
+![image](/assets/images/aut-rollbacks-exec-history-1.png)
+
+If we go to the infrastructure view we can see that the active deployment was rolled back to the previous version:
+
+![image](/assets/images/aut-rollbacks-infra-2.png)
+
+In this case, `V003` was generated by the rollback operation with the contents of a previous deployment. This is the summary of events:
+
+| **Versions before failed deployment**  |  **Versions after failed deployment** |
+|---|---|
+|V001 (nginx:1.14.0)|V002 (nginx:1.15.5-alpine)|
+|V002 (nginx:1.15.5-alpine) → ACTIVE|V003 (nginx:1.14.0) → ACTIVE|
+
+Here we can see one gotcha of the Undo Rollout (Manifest) stage: _**it will rollback a version no matter if the failed pipeline deployed the application or not**_. So make sure that this stage only executes after having a successful deployment of the artifact to rollback, and then some other logic marks the pipeline as failed, otherwise this may have the unintended consequence of rolling back a healthy version. One use case of this may be a pipeline that first deploys some configuration changes to kubernetes, then deploys an updated version of a service. We can have a rollback pipeline execute to rollback the config changes if the main service deployment fails. 
```

Comment:
In "...only executes after having a successful deployment..." remove the "having" (so it's just "...only executes after a successful deployment..."